File: Memento.cpp
#include<cstdio>
#include<iostream>
#include<vector>
#include<stack>
#include<string>
#include<fstream>
#include "ChessBoard.cpp"

using namespace std;

class ChessMemento
{
public:
    int player;
    int w, h;
    vector< vector<char>> board;

    ChessMemento(int w, int h, int player, vector< vector<char>> board){
        this->board = board;
        this->w = w;
        this->h = h;
        this->player = player;
    }


};

class ChessCaretaker
{
public:
    vector<ChessMemento> MementoList;
    string path = "./save/";

    void pushMemento(ChessMemento memento){
        MementoList.push_back(memento);
    }

    ChessMemento topMemento(){
        ChessMemento temp = MementoList[MementoList.size()-1];
        return temp;
    }

    void popMemento(){
        ChessMemento temp = MementoList[MementoList.size()-1];
        MementoList.pop_back();
    }

    int getSize(){
        return MementoList.size();
    }

    bool saveInText(int fileId = 0){
        string file = path+"save"+to_string(fileId)+".txt";
        cout << file << endl;
        ofstream outFile(file);
        if(outFile.is_open()) 
        {   
            for(int i = 0; i < MementoList.size(); i++){
                ChessMemento temp = MementoList[i];
                outFile << temp.player <<endl;
                outFile << temp.h << endl;
                outFile << temp.w << endl;

                for(int j = 0; j < temp.w; j++){
                    for(int k = 0; k < temp.h; k++){
                        outFile << temp.board[j][k] << " ";
                    }
                    outFile << endl;
                }
            }
            outFile.close();
        }else{
            cout << "Save Fail!" << endl;
            return false;
        }
        return true;
    }

    bool restoreFormText(int fileId = 0 ){
        MementoList.clear();
        string file = path+"save"+to_string(fileId)+".txt";
        fstream inFile(file);
        if(inFile.is_open()){
            int w,h,player;
            while(inFile >> player){
                inFile >> h >> w;
                vector< vector<char>> board(w, vector<char>(h));
                for(int i = 0; i < w; i++){
                    for(int j =0; j < h; j++){
                        inFile >> board[i][j];
                    }
                }
                this->pushMemento(ChessMemento(w, h, player, board));
            }
        }else{
            cout << "Read Fail!" << endl;
            return false;
        }
        return true;
    }
    
};

class ChessOriginator
{
public:
    AbstractChessBoard* board_ptr;
    ChessOriginator(AbstractChessBoard* board_ptr){
        this->board_ptr = board_ptr;
    }

    ChessMemento save(){
        return ChessMemento(this->board_ptr->w, this->board_ptr->h, this->board_ptr->player, this->board_ptr->board);
    }

    void restore(ChessMemento memento){
        this->board_ptr->player = memento.player;
        this->board_ptr->h = memento.h;
        this->board_ptr->w = memento.w;
        this->board_ptr->board = memento.board;
    }

};


// int main(){
//     vector<string> nameList = {"a", "b"};
//     AbstractChessBoard* board_ptr = new WeiqiBoard(8,8,nameList);
//     ChessOriginator originator = ChessOriginator(board_ptr);
//     ChessCaretaker caretaker = ChessCaretaker();
//     caretaker.pushMemento(originator.save());
//     board_ptr->step(1,2, true);
//     caretaker.pushMemento(originator.save());
//     board_ptr->step(3,2, true);
//     caretaker.pushMemento(originator.save());
//     caretaker.saveInText();
//     caretaker.restoreFormText();
//     caretaker.saveInText(1);
//     caretaker.popMemento();
//     originator.restore(caretaker.topMemento());
//     caretaker.saveInText(3);
// }

File: Memento.py
from ChessBoard import *
from abc import ABC, abstractmethod
import pickle
import numpy as np
import copy
import os


class ChessMemento:
    def __init__(self, chessBoard) -> None:
        self.chessBoard = copy.deepcopy(chessBoard)
        # self.w = w
        # self.h = h
        # self.player = player


class ChessCaretaker:
    def __init__(self) -> None:
        self.path = "./save/"
        self.MementoList = []

    def pushMemento(self, memento):
        self.MementoList.append(memento)
    
    def topMemento(self):
        temp = self.MementoList[-1]
        return temp

    def getMemento(self, round):
        temp = self.MementoList[round]
        return temp

    def popMemento(self):
        self.MementoList.pop()

    def getSize(self):
        return len(self.MementoList)
    
    def saveInPkl(self, user, fileId = 0) -> bool:
        path = self.path+user+"/"
        if(os.path.exists(path) is False):
            os.makedirs(path)
        file = path+"save{}.pkl".format(fileId)
        try:
            with open(file, 'wb') as f:
                pickle.dump(self.MementoList, f)
                print("存储文件{}成功".format(fileId))
                return True
        except:
            print("存储文件{}失败".format(fileId))
            return False

    def restoreFromPkl(self, user, fileId = 0) -> bool:
        self.MementoList.clear()
        path = self.path+user+"/"
        if(os.path.exists(path) is False):
            os.makedirs(path)
        file = path+"save{}.pkl".format(fileId)
        try:
            with open(file, 'rb') as f:
                self.MementoList = pickle.load(f)
                print("加载文件{}成功".format(fileId))
            return True
        except:
            print("加载文件{}失败".format(fileId))
            return False

class ChessOriginator:
    def __init__(self, chessBoard) -> None:
        self.chessBoard = chessBoard

    def save(self):
        return ChessMemento(self.chessBoard)

    def restore(self, memento):
        self.chessBoard = memento.chessBoard
